### PR TITLE
Cursor screenline public

### PR DESCRIPTION
**Pull request by chrisbra**
As requested in https://github.com/vim/vim/pull/4693#issuecomment-529642042 amend the cursorlineopt setting and allow a 'screenline' value. The value `both` has been aliased to the option `number,line` which is now the new default. To set the value do `:set culopt=screenline` or `:set culopt=screenline,number`. Setting `both,screenline` or `line,screenline` will be an error.

I have been conscious in adjusting the `win_line()` function. It needed quite a bit additional code to handle all different special cases, like when 'list' is set or when a multibyte character does not fit on the screen and gets wrapped and handling are different drawing states. Also, it might cause a small performance penalty, because the caching using `w_last_cursorline` does not work, since we are only highlighting parts of the file.  

BTW: I noticed that the default cursorline highlighting does also underline the `sbr` setting (which is a bit strange, since it does not actually contain to the line content). I almost fixed it, but a test started to fail, so I left it as is (and is only a minor annoyance).

To that end, I added quite a few screen tests, that should make sure that most of the special cases are correctly handled. 

I think most of the special cases are explained in the corresponding commits. I left those commits there and did not squash them to allow for easier review, but I can squash if needed.

### CODE REVIEW

File: src/option.c [hunk 1]
```diff
@@ -9515,3 +9522,55 @@ get_winbuf_options(int bufopt)
     return d;
 }
 #endif
+
+#ifdef FEAT_SYN_HL
+/*
+ * This is called when 'culopt' is change
```
Comment (dpelle) [inline src/option.c]: is change → is changed

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -2468,17 +2468,21 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+	"line"		Highlight the text line of the cursor with
 			CursorLine |hl-CursorLine|.
-	    "number"	Highlight the line number of the cursor with
+	"screenline"	Highlight only the screen line of the cursor with
+			CursorLine |hl-CursorLine|.
+	"number"	Highlight the line number of the cursor with
 			CursorLineNr |hl-CursorLineNr|.
-	    "both"	Highlight as both "line" and "number" are set.
+	Special value:
+	"both"		Alias for the values "line,number".
```
Comment (dpelle) [inline runtime/doc/options.txt:2485]: "both" is a strange name, now that 'culopt' has more than 2 options. But I understand it's needed for backward compatibility.

File: src/screen.c [hunk 3]
```diff
@@ -158,6 +158,9 @@ static void win_redr_custom(win_T *wp, int draw_ruler);
 #ifdef FEAT_CMDL_INFO
 static void win_redr_ruler(win_T *wp, int always, int ignore_pum);
 #endif
+#ifdef FEAT_SYN_HL
+static void margin_columns_win(win_T *wp, int *lmargin, int *rmargin);
+ #endif
```
Comment (dpelle) [inline src/screen.c]: Indentation is wrong here.

File: src/screen.c [hunk 4]
```diff
@@ -5127,6 +5233,11 @@ win_line(
 		    {
 			n_attr = n_extra + 1;
 			extra_attr = HL_ATTR(HLF_8);
+			saved_extra_attr = extra_attr;
+#ifdef FEAT_SYN_HL
+			if (need_cul_screenline)
+			    extra_attr = hl_combine_attr(extra_attr, HL_ATTR(HLF_CUL));
+ #endif
```
Comment (dpelle) [inline src/screen.c]: Indentation not aligned with #ifdef

**Comment by brammool** [comment]
Christian wrote:

> As requested in
> https://github.com/vim/vim/pull/4693#issuecomment-529642042 amend the
> cursorlineopt setting and allow a &#39;screenline&#39; value. The
> value `both` has been aliased to the option `number,line` which is now
> the new default. To set the value do `:set culopt=screenline` or `:set
> culopt=screenline,number`. Setting `both,screenline` or
> `line,screenline` will be an error.
> 
> I have been conscious in adjusting the `win_line()` function. It
> needed quite a bit additional code to handle all different special
> cases, like when &#39;list&#39; is set or when a multibyte character
> does not fit on the screen and gets wrapped and handling are different
> drawing states. Also, it might cause a small performance penalty,
> because the caching using `w_last_cursorline` does not work, since we
> are only highlighting parts of the file.  

I was going to include this, but with a simple test the highlighting is
very wrong.  Try editing src/main.c with:
	hi CursorLine cterm=underline	

With 'cursorlineopt' set to "both" the number column isn't highlighted.
With set to "screenline" lots of highlighting is missing.


-- 
If VIM were a woman, I'd marry her.  Slim, organized, helpful
and beautiful; what's not to like?     --David A. Rogers

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Pull request closed without merging**